# Task 4.9 Part 1

## Contents List:

- Import libraries and customers.csv

- Wrangle the df_cust dataframe
    - Renaming columns

- Data consistency checks
    - Check for missing values, duplicates, and mixed-type data

- Combine df_cust with df_orders_products_merged

- Export orders_products_customers_merged.pkl

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Import customers.csv as dataframe

path = r'C:\Users\susan\OneDrive\Desktop\Data Analytics Program\June 2023 Instacart Basket Analysis'

In [3]:
df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Check output

df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# Check shape

df_cust.shape

(206209, 10)

# Step 4: Wrangle the df_cust dataframe

In [6]:
# Rename 'Surnam' to 'Last Name' for additional clarity

df_cust.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [7]:
# Rename 'STATE' to 'state'

df_cust.rename(columns = {'STATE' : 'State'}, inplace = True)

In [8]:
# Check changes

df_cust.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## Each column of df_cust might be necessary for data analysis later on, so I will not drop any columns for now

# Step 5: Data Consistency Checks

## Check for missing values in df_cust

In [9]:
df_cust.isnull().sum()

user_id             0
First Name      11259
Last Name           0
Gender              0
State               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

## There are 11,259 missing first names, which is 5.46% of the data. Since it is not possible to fill in the missing values without an additional source, I will leave it as missing values which should not affect the results

## Check for duplicates in df_cust

In [10]:
df_dups = df_cust[df_cust.duplicated()]

In [11]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, n_dependants, fam_status, income]
Index: []

## No duplicates found

## Check for mixed-type data in df_cust

In [12]:
for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

First Name


## 'First Name' was found to be a mixed type column. This is because it contains null values as well. I will change the data type to String

In [13]:
df_cust['First Name'] = df_cust['First Name'].astype('str')

In [14]:
# Recheck for mixed-type data in df_cust

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

In [15]:
# Check missing values again after data type change

df_cust.isnull().sum()

user_id         0
First Name      0
Last Name       0
Gender          0
State           0
Age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

## 'First Name' now shows 0 missing values because the data type was changed to string and does not include the null values anymore

# Step 6: Combine df_cust with df_orders_products_merged

In [16]:
# Import df_orders_products_merged dataframe

df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))

In [17]:
# Check output

df_orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day     busiest_days  \
0         Soda  ...  Mid range product  Regularly busy   Regularly busy   
1         Soda  ...  Mid range product  Regularly busy  Least busy days   
2         Soda  ...  Mid range product  Regularly busy  Least busy days   
3         Soda  ...  Mid range product      Least busy  Least busy days   
4         Soda  ...  Mid range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag mean_price spending_flag  \
0        Average orders        10  New customer   6.367797   Low spender   
1        Average orders        10  New customer   6.367797   Low spender   
2           Most orders        10  New customer   6.367797   Low spender   
3        Average orders        10  New customer   6.367797   Low spender   
4           Most orders        10  New customer   6.367797   Low spender   

   median_days_since_prior_order         frequency_flag  
0                           20.5  Non-frequent customer  
1                           20.5  Non-frequent customer  
2                           20.5  Non-frequent customer  
3                           20.5  Non-frequent customer  
4                           20.5  Non-frequent customer  

[5 rows x 24 columns]

In [18]:
# Check shape

df_orders_products_merged.shape

(32404859, 24)

In [19]:
# Mark unusually large price values as missing

df_orders_products_merged.loc[df_orders_products_merged['prices'] > 100, 'prices'] = np.nan

In [20]:
# Check what the max value is of 'prices' column

df_orders_products_merged['prices'].max()

25.0

In [21]:
# Merge df_orders_products_merged with df_cust using 'user_id' column

df_merged = df_orders_products_merged.merge(df_cust, on = 'user_id')

In [22]:
# Check output

df_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...         frequency_flag  First Name  Last Name  Gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     State Age date_joined n_dependants  fam_status income  
0  Alabama  31   2/17/2019            3     married  40423  
1  Alabama  31   2/17/2019            3     married  40423  
2  Alabama  31   2/17/2019            3     married  40423  
3  Alabama  31   2/17/2019            3     married  40423  
4  Alabama  31   2/17/2019            3     married  40423  

[5 rows x 33 columns]

In [23]:
# Check shape

df_merged.shape

(32404859, 33)

In [24]:
# Check basic info of df_merged

df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 33 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   order_id                       int64   
 1   user_id                        int64   
 2   order_number                   int64   
 3   orders_day_of_week             int64   
 4   order_hour_of_day              int64   
 5   days_since_prior_order         float64 
 6   product_id                     int64   
 7   add_to_cart_order              int64   
 8   reordered                      int64   
 9   product_name                   object  
 10  aisle_id                       int64   
 11  department_id                  int64   
 12  prices                         float64 
 13  _merge                         category
 14  price_range_loc                object  
 15  busiest_day                    object  
 16  busiest_days                   object  
 17  busiest_period_of_day    

# Export df_merged

In [25]:
df_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customers_merged.pkl'))